### PR TITLE
Added documentloaded, metadataloaded, and pageremoved API events

### LINKED ISSUE #5765: Add events to the viewer API.
Based on https://github.com/hypothesis/h/blob/document-pdfjs/docs/internals/pdfjs.rst:
- [ ] Need an event when document loaded -- `documentloaded`
- [ ] Need an event when metadata is loaded -- `metadataloaded`
- [ ] Need an event when page is cleared (removed from the cache) -- `pageremoved`

Also add state properties to indicate that those events were already triggered.

/cc @csillag

### PR DESCRIPTION
**Pull request by mbbaig**
Added documentloaded and metadataloaded events

Removed debug code

Fixes #5765

### CODE REVIEW

**Comment by timvandermeij** [comment]
Tip: if your PR fixes an open issue, add the issue to the original description. This way it will be closed automatically by GitHub once the PR is merged. I have just done that for you.

**Comment by timvandermeij** [comment]
I have added some review comments. Once you have fixed these, make sure to also resolve merge conflicts and squash the commits before you push.

**Comment by mbbaig** [comment]
I'll remove the quotes. The attributes are the state properties you mentioned. But they do seem unnecessary to me, and can be removed. 

As fo rthe PR. When I was creating the it say the changes can't be automatically merged. What can cause that?

**Comment by CodingFabian** [comment]
@mbbaig it contains merge conflicts, the easiest way to resolve this is to pull the latest master from this repository into your master and rebase your issue branch.

**Comment by mbbaig** [comment]
@CodingFabian I did that but it still say can't merge

**Comment by CodingFabian** [comment]
@mbbaig nope, the current HEAD of master is 89cf0720262eaa533451b22eb70726c4e4cba6c5
your commit has 6d8adb31ad5c4c9ed44b7ff3b1bff453a9e96411 as parent

**Comment by mbbaig** [comment]
@CodingFabian yeah got it this time. Thanks

**Comment by CodingFabian** [comment]
now you just need to fix the merge :)

**Comment by mbbaig** [comment]
yeah just caught that

File: web/viewer.js [hunk 1]
```diff
@@ -1025,7 +1032,14 @@ var PDFViewerApplication = {
         console.warn('Warning: AcroForm/XFA is not supported');
         self.fallback(PDFJS.UNSUPPORTED_FEATURES.forms);
       }
-
+
```
Comment (timvandermeij) [inline web/viewer.js]: Nit: unnecessary change (only spaces). Please remove this and also try to adjust your editor to remove such whitespace.

File: web/pdf_page_view.js [hunk 2]
```diff
@@ -128,7 +128,15 @@ var PDFPageView = (function PDFPageViewClosure() {
         div.removeChild(node);
       }
       div.removeAttribute('data-loaded');
+      div.setAttribute('data-pageremoved-triggered', true);
```
Comment (timvandermeij) [inline web/pdf_page_view.js]: @yurydelendik Do we need such attributes? It seems unnecessary to me as one can simply wait until the event fires...

Comment (mbbaig) [inline web/pdf_page_view.js]: I change the following
```javascript
detail: { triggered: true },
```
to something like:
```javascript
detail: { triggered: div.getAttribute('data-page-number')},
```
so the event itself has the page number of the page the event fired on. Then we no longer need the attributes but still have useful information.

Comment (yurydelendik) [inline web/pdf_page_view.js]: We don't need this attribute. The page renderingState will show if rendered page is present.

File: web/pdf_page_view.js [hunk 3]
```diff
@@ -128,7 +128,15 @@ var PDFPageView = (function PDFPageViewClosure() {
...
 
+      var event = new CustomEvent('pageremoved', {
+        detail: { triggered: true },
```
Comment (yurydelendik) [inline web/pdf_page_view.js]: Instead of `triggered` you can use `pageNumber` here.

File: web/viewer.js [hunk 4]
```diff
@@ -892,6 +892,13 @@ var PDFViewerApplication = {
         PDFHistory.initialize(self.documentFingerprint, self);
       }
 
+      var docLoadedEvent = new CustomEvent('documentloaded', {
+        detail: { triggered: true },
```
Comment (yurydelendik) [inline web/viewer.js]: The same here and below -- you don't have to specify `triggered`.

**Comment by mbbaig** [comment]
Made the required changes

File: web/pdf_page_view.js [hunk 5]
```diff
@@ -129,6 +129,13 @@ var PDFPageView = (function PDFPageViewClosure() {
       }
       div.removeAttribute('data-loaded');
 
+      var event = new CustomEvent('pageremoved', {
+        detail: { pageNumber: div.getAttribute('data-page-number') },
```
Comment (Snuffleupagus) [inline web/pdf_page_view.js]: Is there a reason that you're not just using `pageNumber: this.id` here?

Comment (mbbaig) [inline web/pdf_page_view.js]: Yes. The reason being that I completely forgot about ```this.id``` :)

I changed that locally.

**Comment by mbbaig** [comment]
Haven't pushed the change yet. Waiting for other input.

**Comment by timvandermeij** [comment]
Aside from the above comment, this looks good to me.

File: web/viewer.js [hunk 6]
```diff
@@ -892,6 +892,12 @@ var PDFViewerApplication = {
         PDFHistory.initialize(self.documentFingerprint, self);
       }
 
+      var docLoadedEvent = new CustomEvent('documentloaded', {
```
Comment (brendandahl) [inline web/viewer.js]: This is confusing to have because we also have a documentload event.

Comment (brendandahl) [inline web/viewer.js]: 'firstpageready' is more accurate for what this event is

File: web/viewer.js [hunk 7]
```diff
@@ -1026,6 +1032,12 @@ var PDFViewerApplication = {
         self.fallback(PDFJS.UNSUPPORTED_FEATURES.forms);
       }
 
+      var metaLoadedEvent = new CustomEvent('metadataloaded', {
```
Comment (brendandahl) [inline web/viewer.js]: drop the 'ed' in 'medatadataloaded' so we're consistent with documentload

**Comment by brendandahl** [comment]
Disregard the above for now. We need to flesh out what event we want and what they really mean.

**Comment by meltuhamy** [comment]
Any updates regarding these events? @mbbaig looks like there's conflicts again. Is this branch dead?

**Comment by mbbaig** [comment]
I'll take a look at the conflicts

**Comment by brendandahl** [comment]
Sorry for the delay in responding to this.  It's hard to decide where these events should go and be called because I can't read the original context where this was requested. This is my current thinking:

- metadataloaded is in a good spot and doesn't seem misleading. Keep it.
- documentloaded is misleading since the 'documentloaded' event could happen before the 'documentload' event and it's unclear what the real difference is by there names. Until I hear of an actual use case for the 'documentloaded' event vs 'documentload' event I'd like to remove the 'documentloaded' event from this PR
- 'pageremoved' is also a bit misleading because it's not when the page is actually destroyed or removed, just 'reset' so let's just be consistent and use the function name and call this 'pagereset'

Address the two comments above and this should be good to go.

**Comment by timvandermeij** [comment]
The event code changed quite a lot recently to make use of an event bus. As a consequence, this PR needs to be rewritten. Since this PR has been stale for while now and has not been squared/rebased, I'm going to close this PR. We keep on tracking the addition of the events in #5765. If someone is willing to create a new patch for this using the event bus code we have now, please open a new PR for that. Thanks.
